### ISSUE TITLE
reverse order of items in deck

### ISSUE THREAD

**Issue by smcoll** [yourlabs/django-autocomplete-light#331]
i'm using django-sortedm2m to handle an m2m with ordering.  However, the MultipleChoiceWidget is rendering the related items in reverse order (opposite order than the hidden select) for pre-existing values.  How can i reverse the ordering of the initial values in the deck so that they mirror the ordering of the hidden select?

i don't mean to order the queryset of the autocomplete.

**Comment by smcoll** [comment]
Further investigation shows that my "reverse" observation is coincidental; it's just a different ordering.

```
# models.py
class Bar(models.Model):
    name = models.CharField(max_length=20)

class Foo(models.Model):
    bars = SortedManyToManyField(Bar)

# admin.py
class FooAdmin(admin.ModelAdmin):
    form = autocomplete_light.modelform_factory(Foo)
    # raw_id_fields = ('bars',)
    model = Foo
```

In the console, if i return an ordered m2m queryset, it has this order:

```
>>> foo = Foo.objects.first()
>>> foo.bars.values_list('id', flat=True)
[15,11,10]
```

But the widget has them in a different order (i think it reflects the default ordering of Bar, rather than the ordering of the m2m queryset, which is determined by a sort_order field on the intermediate model):
```
<select style="display:none" class="value-select" name="bars" id="id_bars" multiple="multiple">
  <option value="11" selected="selected">11</option>
  <option value="10" selected="selected">10</option>
  <option value="15" selected="selected">15</option>
</select>
```

If i switch to `raw_id_fields` (in lieu of the form), the ordering is correct again: `15,11,10`.

**Comment by smcoll** [comment]
Perhaps this is related to #360

**Comment by smcoll** [comment]
Found it!  It's happening in the `render()` method of `autocomplete_light.WidgetBase`.  The `value` argument has the correct ordering, but `autocomplete.choices_for_values()` returns a corresponding list in an order matching its own queryset, not the original ordering.  See the commented code below:

```
class WidgetBase(object):
    # ...

    def render(self, name, value, attrs=None):
        widget_attrs = self.build_widget_attrs(name)
        attrs = self.build_attrs(attrs)
        self.html_id = attrs.pop('id', name)

        # value = [15,11,10]

        autocomplete = self.autocomplete(values=value)
        # autocomplete.choices is a queryset of it's own ordering
        # autocomplete.choices = [12,13,14,11,10,9,3,16,17,15,18]

        choices = autocomplete.choices_for_values()
        # choices has the wrong order now
        # choices.values_list('id', flat=True) = [11,10,15]

        values = [autocomplete.choice_value(c) for c in choices]
        # now, values is the same list as value, but in the wrong order
        # values = [11,10,15]
        
        # ...
```

@jpic do you have a preference for where the fix might go?  In `autocomplete.choices_for_values()`, or as a new list comprehension where `values` is assigned?

**Comment by smcoll** [comment]
@jpic i opted, for the time being, to do this:

```
        values = [c for c in value if c in choices.all().values_list('id', flat=True)]
```

Would that break anything?  Is there any purpose in constructing values aside form ensuring that each value exists in the autocomplete's queryset (choices)?

That solves the order of the options in the select, but now the deck is still building in its own order.  Perhaps you can provide more context about that.

**Comment by jpic** [comment]
It looks like it would make sense to improve choices_for_values respect the
order of values just in case, that won't break the case where order is
irrelevant will it?  ;)

**Comment by smcoll** [comment]
@jpic it may.  `AutocompleteModel.choices_for_values` explicitly calls `AutocompleteModel.order_choices`, which is where the re-ordering happens.  i presume the reordering is desired in some cases, otherwise the method would be called.

Do you know why `order_choices` is necessary in that context?  Wouldn't a related queryset already have the appropriate ordering in that case?

**Comment by smcoll** [comment]
@jpic perhaps `choices_for_values` could have a `preserve_ordering=False` argument that could be leveraged in some contexts?

**Comment by jpic** [comment]
Order should be respected by autocomplete-light.

Next release is planned in ~2-3 months if you want to get your patch in.

**Comment by smcoll** [comment]
@jpic can you shed some light on why `AutocompleteModel.choices_for_values` explicitly calls `AutocompleteModel.order_choices`, rather than respecting the original order?

For the record, my final implementation works for me but doesn't look like a good candidate for a PR:

```
class WidgetBase(object):
    # ...
    def render(self, name, value, attrs=None):
        widget_attrs = self.build_widget_attrs(name)
        attrs = self.build_attrs(attrs)
        self.html_id = attrs.pop('id', name)

        autocomplete = self.autocomplete(values=value)
        choices = autocomplete.choices_for_values()

        # restore order of choices from value
        # inspired by http://blog.mathieu-leplatre.info/django-create-a-queryset-from-a-list-preserving-order.html
        if value:
            clauses = ' '.join(['WHEN id=%s THEN %s' % (pk, i) for i, pk in enumerate(value)])
            ordering = 'CASE %s END' % clauses
            choices = choices.extra(
                    select={'ordering': ordering}, order_by=('ordering',))
        values = [autocomplete.choice_value(c) for c in choices]

        context = {
            'name': name,
            'values': values,
            'choices': choices,
            'widget': self,
            'attrs': safestring.mark_safe(flatatt(attrs)),
            'widget_attrs': safestring.mark_safe(flatatt(widget_attrs)),
            'autocomplete': autocomplete,
        }
        context.update(self.extra_context)

        template = getattr(autocomplete, 'widget_template',
                self.widget_template)
        return safestring.mark_safe(render_to_string(self.widget_template, context))
```

**Comment by jpic** [comment]
It would be a good idea to compare different implementation options, so we need a test case first.

You can run tests by running this command in the django-autocomplete-light repo `tox -e py34-django18`.

**Comment by jpic** [comment]
I think this part is not very efficient:

        choices = autocomplete.choices_for_values()
        values = [autocomplete.choice_value(c) for c in choices]

Maybe it would be better to have something like this instead:

        choices, values = autocomplete.choices_values()

Lemme know what you think

**Comment by smcoll** [comment]
@jpic added #407, which just has a failing test.

i'm thinking that for starters, `AutocompleteModel.choices_for_values()` needn't use `order_choices()`; the choices need to be ordered in `choices_for_request()` so the deck can be ordered according to configuration, but `choices_for_values()` (or something analogous), being used in the widget, ought to respect the ordering given by the queryset.  Is that what you're thinking?

**Comment by blueyed** [comment]
@smcoll 
:+1: Makes sense to me.

**Comment by jpic** [comment]
Fixed in 2.1.0rc4.

Could you please test it and let us know ?

This is the final 2.1.0 release blocker ^^

**Comment by smcoll** [comment]
@jpic confirmed!  Thanks very much.
